### PR TITLE
[clang][NFC] Use range-based for loops

### PR DESCRIPTION
**Pull request by MagentaTreehouse**
Use range-based for loops. In addition, extracted a loop from `CXXRecordDecl::completeDefinition` to eliminate the `Done` flag, and only construct `MyFinalOverriders` when `FinalOverriders` is null.

### CODE REVIEW

File: clang/lib/AST/DeclCXX.cpp [hunk 1]
```diff
@@ -2056,40 +2056,40 @@ void CXXRecordDecl::completeDefinition() {
   completeDefinition(nullptr);
 }
 
+static bool hasPureVirtualFinalOverrider(
+    const CXXRecordDecl &RD, const CXXFinalOverriderMap *FinalOverriders) {
+  auto ExistsIn = [](const CXXFinalOverriderMap &FinalOverriders) {
+    for (const auto &[_, M] : FinalOverriders) {
+      for (const auto &[_, SO] : M) {
```
Comment (Endilll) [inline clang/lib/AST/DeclCXX.cpp:2063]: While we're at it, can we use more descriptive names than `M` and `SO`? I'm not even sure what the latter means. I also have reservations towards `auto` here.

**Review by Endilll — COMMENTED** [review]
This looks good overall, but I have minor suggestions.

File: clang/lib/AST/DeclCXX.cpp (hunk 1 above)
Comment (Sirraide) [reply clang/lib/AST/DeclCXX.cpp:2063]: Yeah, maybe not using structured bindings here would make this more legible...

Comment (MagentaTreehouse) [reply clang/lib/AST/DeclCXX.cpp]: Updated the code with more descriptive names and removed structured bindings for the outer loop.

Regarding the inner loop, I found that the type of the loop variable cannot be spelled with the public APIs of the class `OverridingMethods` we are iterating through, so I kept that structured binding. But, its readability improved with the updated name. Please let me know if you have further suggestions.

**Review by Endilll — APPROVED** [review]

File: clang/lib/AST/DeclCXX.cpp [hunk 2]
```diff
@@ -2056,40 +2056,41 @@ void CXXRecordDecl::completeDefinition() {
...
+      }
+    }
+    return false;
+  };
+
+  if (FinalOverriders)
+    return ExistsIn(*FinalOverriders);
+  CXXFinalOverriderMap MyFinalOverriders;
+  RD.getFinalOverriders(MyFinalOverriders);
+  return ExistsIn(MyFinalOverriders);
```
Comment (MitalAshok) [inline clang/lib/AST/DeclCXX.cpp:2079]: Thoughts on moving this to the front like:

```c++
static bool hasPureVirtualFinalOverrider(
    const CXXRecordDecl &RD, const CXXFinalOverriderMap *FinalOverriders) {
  if (!FinalOverriders) {
    CXXFinalOverriderMap MyFinalOverriders;
    RD.getFinalOverriders(MyFinalOverriders);
    return hasPureVirtualFinalOverrider(RD, &MyFinalOverriders);
  }

  for (/*...*/) {
    // ...
  }
  return false;
}
```

?

This also gets rid of the lambda

Comment (MagentaTreehouse) [reply clang/lib/AST/DeclCXX.cpp]: Done

**Review by MitalAshok — APPROVED** [review]
LGTM

**Pull request merged**